ArcGISRuntimeSDKQt_CppSamples/Routing/OfflineRouting/OfflineRouting.cpp

```diff
@@ -0,0 +1,271 @@
+// [WriteFile Name=OfflineRouting, Category=Analysis]
+// [Legal]
+// Copyright 2020 Esri.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// [Legal]
+
+#ifdef PCH_BUILD
+#include "pch.hpp"
+#endif // PCH_BUILD
+
+#include "OfflineRouting.h"
+
+#include "ArcGISTiledLayer.h"
+#include "Envelope.h"
+#include "Graphic.h"
+#include "GraphicsOverlay.h"
+#include "Map.h"
+#include "MapQuickView.h"
+#include "Polyline.h"
+#include "RouteParameters.h"
+#include "RouteResult.h"
+#include "RouteTask.h"
+#include "SimpleMarkerSymbol.h"
+#include "SimpleLineSymbol.h"
+#include "SimpleRenderer.h"
+#include "Stop.h"
+#include "TextSymbol.h"
+#include "TileCache.h"
+
+#include <QDir>
+#include <QScopedPointer>
+
+using namespace Esri::ArcGISRuntime;
+
+// helper method to get cross platform data path
+namespace
+{
+QString defaultDataPath()
+{
+  QString dataPath;
+
+#ifdef Q_OS_ANDROID
+  dataPath = "/sdcard";
+#elif defined Q_OS_IOS
+  dataPath = QStandardPaths::writableLocation(QStandardPaths::DocumentsLocation);
+#else
+  dataPath = QDir::homePath();
+#endif
+
+  return dataPath;
+}
+
+} // namespace
+
+OfflineRouting::OfflineRouting(QObject* parent /* = nullptr */):
+  QObject(parent)
+{
+  const QString fileLocation = QString("%1/ArcGIS/Runtime/Data/tpk/san_diego/streetmap_SD.tpk").arg(defaultDataPath());
```

Comment:
optional - if user's don't download the data 1st the file worn't exist We could add a check for that which fails gracefully